File: src/cpp/sum_thread.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <chrono>
#include <random>
#include <numeric>
#include <string>

class SumThread {
private:
    const std::vector<int>& numbers;
    size_t start_index;
    size_t end_index;
    long long partial_sum;

public:
    SumThread(const std::vector<int>& nums, size_t start, size_t end)
        : numbers(nums), start_index(start), end_index(end), partial_sum(0) {}

    void calculate() {
        partial_sum = std::accumulate(numbers.begin() + start_index, 
                                    numbers.begin() + end_index, 0LL);
    }

    long long getPartialSum() const { return partial_sum; }
};

std::pair<long long, double> parallel_sum(const std::vector<int>& numbers, int num_threads) {
    auto start_time = std::chrono::high_resolution_clock::now();
    
    std::vector<std::thread> threads;
    std::vector<SumThread> sum_threads;
    
    size_t chunk_size = numbers.size() / num_threads;
    
    // Create and start threads
    for (int i = 0; i < num_threads; ++i) {
        size_t start_idx = i * chunk_size;
        size_t end_idx = (i == num_threads - 1) ? numbers.size() : (i + 1) * chunk_size;
        
        sum_threads.emplace_back(numbers, start_idx, end_idx);
        threads.emplace_back(&SumThread::calculate, &sum_threads.back());
    }
    
    // Wait for all threads
    for (auto& thread : threads) {
        thread.join();
    }
    
    // Calculate total sum
    long long total_sum = 0;
    for (const auto& sum_thread : sum_threads) {
        total_sum += sum_thread.getPartialSum();
    }
    
    auto end_time = std::chrono::high_resolution_clock::now();
    double duration = std::chrono::duration<double>(end_time - start_time).count();
    
    return {total_sum, duration};
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <size> <num_threads>\n";
        return 1;
    }

    int size = std::stoi(argv[1]);
    int num_threads = std::stoi(argv[2]);

    // Test cases
    std::vector<std::vector<int>> test_cases;
    
    // Case 1: Sequential numbers
    std::vector<int> seq_numbers(size);
    std::iota(seq_numbers.begin(), seq_numbers.end(), 0);
    test_cases.push_back(seq_numbers);
    
    // Case 2: Random numbers
    std::vector<int> random_numbers(size);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(1, 100);
    for (auto& num : random_numbers) {
        num = dis(gen);
    }
    test_cases.push_back(random_numbers);
    
    // Case 3: Uniform numbers
    std::vector<int> uniform_numbers(size, 1);
    test_cases.push_back(uniform_numbers);
    
    for (size_t i = 0; i < test_cases.size(); ++i) {
        std::cout << "\nTest Case " << i + 1 << ":\n";
        std::cout << "List size: " << test_cases[i].size() << "\n";
        
        // Regular sum
        auto start = std::chrono::high_resolution_clock::now();
        long long regular_sum = std::accumulate(test_cases[i].begin(), 
                                              test_cases[i].end(), 0LL);
        auto end = std::chrono::high_resolution_clock::now();
        double regular_time = std::chrono::duration<double>(end - start).count();
        
        std::cout << "Regular sum: " << regular_sum << "\n";
        std::cout << "Regular time: " << regular_time << " seconds\n";
        
        // Parallel sum
        auto [parallel_result, parallel_time] = parallel_sum(test_cases[i], num_threads);
        std::cout << "Parallel sum: " << parallel_result << "\n";
        std::cout << "Parallel time: " << parallel_time << " seconds\n";
        std::cout << "Speed improvement: " << (regular_time/parallel_time) << "x\n";
    }
    
    return 0;
}

File: src/python/sum_thread.py
import threading
import time
import random
import sys

class SumThread(threading.Thread):
    """Thread class to calculate partial sum of a list"""
    def __init__(self, numbers, start_index, end_index):
        threading.Thread.__init__(self)
        self.numbers = numbers
        self.start_index = start_index
        self.end_index = end_index
        self.partial_sum = 0

    def run(self):
        """Calculate the sum of the assigned portion of the list"""
        self.partial_sum = sum(self.numbers[self.start_index:self.end_index])

def parallel_sum(numbers, num_threads=2):
    """
    Calculate the sum of a list using multiple threads
    
    Args:
        numbers (list): List of numbers to sum
        num_threads (int): Number of threads to use
        
    Returns:
        tuple: Total sum and execution time
    """
    # Record start time
    start_time = time.time()
    
    # Calculate chunk size for each thread
    chunk_size = len(numbers) // num_threads
    
    # Create and start threads
    threads = []
    for i in range(num_threads):
        start_idx = i * chunk_size
        # Handle the last chunk which might be larger
        end_idx = len(numbers) if i == num_threads - 1 else (i + 1) * chunk_size
        
        thread = SumThread(numbers, start_idx, end_idx)
        threads.append(thread)
        thread.start()
    
    # Wait for all threads to complete
    for thread in threads:
        thread.join()
    
    # Calculate total sum from partial sums
    total_sum = sum(thread.partial_sum for thread in threads)
    
    # Calculate execution time
    execution_time = time.time() - start_time
    
    return total_sum, execution_time

def main():
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <size> <num_threads>")
        sys.exit(1)

    size = int(sys.argv[1])
    num_threads = int(sys.argv[2])

    # Test cases
    test_cases = [
        list(range(size)),  # Sequential numbers
        [random.randint(1, 100) for _ in range(size)],  # Random numbers
        [1] * size  # Uniform numbers
    ]
    
    for i, numbers in enumerate(test_cases, 1):
        print(f"\nTest Case {i}:")
        print(f"List size: {len(numbers)}")
        
        # Calculate sum using regular method for comparison
        start_time = time.time()
        regular_sum = sum(numbers)
        regular_time = time.time() - start_time
        print(f"Regular sum: {regular_sum}")
        print(f"Regular execution time: {regular_time:.4f} seconds")
        
        # Calculate sum using parallel method
        parallel_result, parallel_time = parallel_sum(numbers)
        print(f"Parallel sum: {parallel_result}")
        print(f"Parallel execution time: {parallel_time:.4f} seconds")
        print(f"Speed improvement: {(regular_time/parallel_time):.2f}x")

if __name__ == "__main__":
    main()